core/thread/key.go

```diff
@@ -0,0 +1,122 @@
+package thread
+
+import (
+	"fmt"
+
+	mbase "github.com/multiformats/go-multibase"
+	sym "github.com/textileio/go-threads/crypto/symmetric"
+)
+
+var (
+	// Indicates an invalid byte slice was given to KeyFromBytes.
+	ErrInvalidKey = fmt.Errorf("invalid key")
+)
+
+// Key is a thread encryption key with two components.
+// Service key is used to encrypt outer log record linkages.
+// Read key is used to encrypt inner record events.
+type Key struct {
+	sk *sym.Key
+	rk *sym.Key
+}
+
+// NewKey wraps service and read keys.
+func NewKey(sk, rk *sym.Key) Key {
+	if sk == nil {
+		panic("service-key must not be nil")
+	}
+	return Key{sk: sk, rk: rk}
+}
+
+// NewServiceKey wraps a service-only key.
+func NewServiceKey(sk *sym.Key) Key {
+	return Key{sk: sk}
+}
+
+// NewRandomKey returns a random key, which includes a service and read key.
+func NewRandomKey() Key {
+	return Key{sk: sym.New(), rk: sym.New()}
+}
+
+// NewRandomServiceKey returns a random service-only key.
+func NewRandomServiceKey() Key {
+	return Key{sk: sym.New()}
+}
+
+// KeyFromBytes returns a key by wrapping k.
+func KeyFromBytes(b []byte) (k Key, err error) {
+	if len(b) != sym.KeyBytes && len(b) != sym.KeyBytes*2 {
+		return k, ErrInvalidKey
+	}
+	sk, err := sym.FromBytes(b[:sym.KeyBytes])
+	if err != nil {
+		return k, err
+	}
+	var rk *sym.Key
+	if len(b) == sym.KeyBytes*2 {
+		rk, err = sym.FromBytes(b[sym.KeyBytes:])
+		if err != nil {
+			return k, err
+		}
+	}
+	return Key{sk: sk, rk: rk}, nil
+}
+
+// KeyFromString returns a key by decoding a base58-encoded string.
+func KeyFromString(s string) (k Key, err error) {
+	_, b, err := mbase.Decode(s)
+	if err != nil {
+		return k, err
+	}
+	return KeyFromBytes(b)
+}
+
+// Service returns the service key.
+func (k Key) Service() *sym.Key {
+	return k.sk
+}
+
+// Read returns the read key.
+func (k Key) Read() *sym.Key {
+	return k.rk
+}
+
+// Defined returns whether or not key has any defined components.
+// Since it's not possible to have a read key w/o a service key,
+// we just need to check service key.
+func (k Key) Defined() bool {
+	return k.sk != nil
+}
+
+// CanRead returns whether or not read key is available.
+func (k Key) CanRead() bool {
+	return k.rk != nil
+}
+
+// Marshal returns raw key bytes while conforming to Marshaler.
+func (k Key) Marshal() ([]byte, error) {
```

Comment:
The `Marshaler` interface mentioned in the comment [has one method](https://golang.org/pkg/encoding/json/#Marshaler) `MarshalJSON`, not `Marshal`. Also, `Marshaler` outputs json bytes which isn't our case.

Maybe calling this `MarshalBinary` to let `Key` implement [BinaryMarshaler](https://golang.org/pkg/encoding/#BinaryMarshaler), which is a raw binary marshaler?

Reply:
Nice, will do!